### PR TITLE
Fix #36 Block2 request does not seem to notify CoapHandler#onError if…

### LINKED ISSUE #36: Block2 request does not seem to notify CoapHandler#onError if exchange stucks on non-first request.
Hello!
I've been playing around block transfer and noticed one thing.
I use Californium 1.0.3.
I have a Client which sends CON GET and a Server which responds with large payload.
Then, I tried to emulate network loss and added a connector to the Client which skipped some of ACKs.

When I skip the first ACK, after retries and some time it logs "Timeout: retransmission limit reached, exchange failed, message: ..." and my CoapHandler#onError is being invoked.

When I skip any intermediate ACK, it logs the same line but does not invoke CoapHandler#onError.
I've made some debugging.

org.eclipse.californium.core.network.stack.ReliabilityLayer.java:

```
 protected abstract class RetransmissionTask implements Runnable {
	<skimmed>
359:	} else {
	    LOGGER.fine("Timeout: retransmission limit reached, exchange failed, message: "+message);
	        exchange.setTimedOut();
	        message.setTimedOut(true);
	<skimmed>
```

And if the ```message``` is an intermediate block transfer request, it does not have any handlers, so no onError could be called. But ```exchange#request``` has the original request which has its' handlers.

Maybe the ```exchange#request.setTimedOut(true)``` should be called there in the case of block transfer? It looks reasonable for me because in case of success CoapHandler#onLoad of the original request is called.

The sample code will be provided if requested.

### PR DESCRIPTION
**Pull request by evgeniynickolaev**
To: https://github.com/eclipse/californium/issues/36

Signed-off-by: Eugene Nikolaev <eug.nikolaev@gmail.com>

### CODE REVIEW

File: californium-core/src/main/java/org/eclipse/californium/core/network/stack/ReliabilityLayer.java [hunk 1]
```diff
@@ -364,6 +365,15 @@ public void run() {
 					LOGGER.log(Level.FINE, "Timeout: retransmission limit reached, exchange failed, message: {0}", message);
 					exchange.setTimedOut();
 					message.setTimedOut(true);
+					OptionSet options = message.getOptions();
+					boolean isBlockwise = options.hasBlock1() || options.hasBlock2();
+					if (message instanceof Request && isBlockwise) {
```
Comment (sophokles73) [inline californium-core/src/main/java/org/eclipse/californium/core/network/stack/ReliabilityLayer.java]: We should better use `Exchange.origin` field to determine if this is an outbound request. For that purpose we could e.g. add a `Exchange.isOfLocalOrigin()` method.

File: californium-core/src/main/java/org/eclipse/californium/core/network/stack/ReliabilityLayer.java [hunk 2]
```diff
@@ -364,6 +365,15 @@ public void run() {
...
+					    Request request = exchange.getRequest();
+					    if (request != null && request.getMID() != message.getMID()) {
+					        // Notifying the original request about timeout
+					        request.setTimedOut(true);
+					    }
```
Comment (sophokles73) [inline californium-core/src/main/java/org/eclipse/californium/core/network/stack/ReliabilityLayer.java]: I am not sure if we should invoke `request.setTimedOut()` on the original request *in addition* to `message.setTimedOut()` or *instead* of it ...
In the former case we might end up with two listeners being invoked. But maybe that's what is intended if somebody registers a `MessageListener` on the *intermediary* GET of the blockwise transfer.

Comment (evgeniynickolaev) [inline californium-core/src/main/java/org/eclipse/californium/core/network/stack/ReliabilityLayer.java]: It is hard for me to figure out what could be the intent to make a listener for an intermediate request, but your last point looks very reasonable, I think such behavior seems as more expected by a user.

File: californium-core/src/main/java/org/eclipse/californium/core/network/stack/ReliabilityLayer.java (hunk 1 above)
Comment (evgeniynickolaev) [inline californium-core/src/main/java/org/eclipse/californium/core/network/stack/ReliabilityLayer.java]: Great, thanks! What is the best approach to change it - to add a new PR or to add the next commit to this?

Comment (sophokles73) [inline californium-core/src/main/java/org/eclipse/californium/core/network/stack/ReliabilityLayer.java]: Simply add a new commit to this PR

File: californium-core/src/main/java/org/eclipse/californium/core/network/stack/ReliabilityLayer.java (hunk 2 above)
Comment (sophokles73) [inline californium-core/src/main/java/org/eclipse/californium/core/network/stack/ReliabilityLayer.java]: Yes, and it is actually not that easy to get a grip on such an intermediary request from user code (usually these intermediary reuqests are created and issued by Californium). However, a seasoned user can also do blockwise transfer *manually*, i.e. using a sequence of manually crafted and sent GET requests to perform the transfer.

**Comment by evgeniynickolaev** [comment]
Applied review comments

File: californium-core/src/main/java/org/eclipse/californium/core/network/Exchange.java [hunk 3]
```diff
@@ -203,6 +203,10 @@ public Origin getOrigin() {
 		return origin;
 	}
 
+    public boolean isOfLocalOrigin() {
+        return origin == Origin.LOCAL;
+    }
+
```
Comment (sophokles73) [inline californium-core/src/main/java/org/eclipse/californium/core/network/Exchange.java]: Indentation?
You can use the eclipse formatter rules contained in the parent project if you are using eclipse ...

File: californium-core/src/main/java/org/eclipse/californium/core/network/stack/ReliabilityLayer.java [hunk 4]
```diff
@@ -367,10 +367,10 @@ public void run() {
 					message.setTimedOut(true);
 					OptionSet options = message.getOptions();
 					boolean isBlockwise = options.hasBlock1() || options.hasBlock2();
-					if (message instanceof Request && isBlockwise) {
+					if (exchange.isOfLocalOrigin() && isBlockwise) {
```
Comment (sophokles73) [inline californium-core/src/main/java/org/eclipse/californium/core/network/stack/ReliabilityLayer.java]: looks good :+1:

File: californium-core/src/main/java/org/eclipse/californium/core/network/Exchange.java (hunk 3 above)
Comment (evgeniynickolaev) [inline californium-core/src/main/java/org/eclipse/californium/core/network/Exchange.java]: Oh, sorry, will make a commit again

Comment (evgeniynickolaev) [inline californium-core/src/main/java/org/eclipse/californium/core/network/Exchange.java]: Yes, thank you, will use the rules. Just used to use spaces at prev job)

**Pull request merged**